modules/renter/downloadcache.go

```diff
@@ -4,72 +4,152 @@ package renter
-func (udc *unfinishedDownloadChunk) addChunkToCache(data []byte) {
-	if udc.download.staticDestinationType != destinationTypeSeekStream {
+func (dcc *downloadChunkCache) add(data []byte, cacheID string, destinationType string) {
+	if destinationType != destinationTypeSeekStream {
 		// We only cache streaming chunks since browsers and media players tend
 		// to only request a few kib at once when streaming data. That way we can
 		// prevent scheduling the same chunk for download over and over.
 		return
 	}
-	udc.cacheMu.Lock()
-	defer udc.cacheMu.Unlock()
+	dcc.mu.Lock()
+	defer dcc.mu.Unlock()
 
 	// Prune cache if necessary.
-	for len(udc.chunkCache) >= downloadCacheSize {
-		var oldestKey string
-		oldestTime := time.Now()
-
-		// TODO: turn this from a structure where you loop over every element
-		// (O(n) per access) to a min heap (O(log n) per access).
-		// currently not a issue due to cache size remaining small (<20)
-		for id, chunk := range udc.chunkCache {
-			if chunk.lastAccess.Before(oldestTime) {
-				oldestTime = chunk.lastAccess
-				oldestKey = id
-			}
+	if dcc.cacheSize == 0 {
+		build.Critical("DownloadCacheSize is set to zero")
+		return
+	}
+
+	for len(dcc.chunkCacheMap) >= int(dcc.cacheSize) {
+		// Remove from Heap
+		cd := heap.Pop(&dcc.chunkCacheHeap).(*chunkData)
+
+		// Remove from Map
+		if _, ok := dcc.chunkCacheMap[cd.id]; !ok {
+			build.Critical("Cache Data chunk not found in chunkCacheMap.")
 		}
-		delete(udc.chunkCache, oldestKey)
+		delete(dcc.chunkCacheMap, cd.id)
 	}
 
-	udc.chunkCache[udc.staticCacheID] = &cacheData{
+	// Add chunk to Map and Heap
+	cd := &chunkData{
+		id:         cacheID,
 		data:       data,
 		lastAccess: time.Now(),
 	}
+	dcc.chunkCacheMap[cacheID] = cd
+	heap.Push(&dcc.chunkCacheHeap, cd)
+	dcc.chunkCacheHeap.update(cd, cd.id, cd.data, cd.lastAccess)
+}
+
+// init initializes the downloadChunkCache
+func (dcc *downloadChunkCache) init() {
+	dcc.chunkCacheMap = make(map[string]*chunkData)
+	dcc.chunkCacheHeap = make(chunkCacheHeap, 0, defaultDownloadCacheSize)
+	dcc.cacheSize = defaultDownloadCacheSize
+	heap.Init(&dcc.chunkCacheHeap)
 }
 
-// managedTryCache tries to retrieve the chunk from the renter's cache. If
+// retrieve tries to retrieve the chunk from the renter's cache. If
 // successful it will write the data to the destination and stop the download
 // if it was the last missing chunk. The function returns true if the chunk was
 // in the cache.
-// TODO in the future we might need cache invalidation. At the
+// Using the entire unfisihedDownloadChunk as the argument as there are seven different fields
+// used from unfinishedDownloadChunk and it allows using udc.fail()
+//
+// TODO: in the future we might need cache invalidation. At the
 // moment this doesn't worry us since our files are static.
-func (r *Renter) managedTryCache(udc *unfinishedDownloadChunk) bool {
+func (dcc *downloadChunkCache) retrieve(udc *unfinishedDownloadChunk) bool {
```

Comment:
This needs to be `Retrieve`. That way it is clear for a caller that no lock can be held while calling `Retrieve` and that the `chunkCache` will handle locking itself.